### PR TITLE
Adding the ability to route all non-file requests to the index file for SPAs.

### PR DESCRIPTION
**Pull request by mmcdermott**
Also adds a custom index file option to the CLI.

### CODE REVIEW

**Comment by dizel3d** [comment]
The PR covers issue #175

File: bin/cli.js [hunk 1]
```diff
@@ -92,20 +101,33 @@ if (argv.gzip){
...
+            file.serveFile(argv['indexFile'], 500, {}, request, response);
+        } else {
+            file.serve(request, response, callback);
+        }
     }).resume();
 }).listen(+argv.port, argv['host-address']);
 
 console.log('serving "' + dir + '" at http://' + argv['host-address'] + ':' + argv.port);
+if (argv.spa) {
+  console.log('serving as a single page app (all non-file requests redirect to ' + argv['indexFile'] +')')
```
Comment (cloudhead) [inline bin/cli.js]: `;`

File: bin/cli.js [hunk 2]
```diff
@@ -92,20 +101,33 @@ if (argv.gzip){
...
-        });
+        var callback = function(e, rsp) {
+          if (e && e.status === 404) {
+              response.writeHead(e.status, e.headers);
+              response.end("Not Found");
+              log(request, response);
+          } else {
+              log(request, response);
+          }
+        }
```
Comment (cloudhead) [inline bin/cli.js]: `;`

File: bin/cli.js [hunk 3]
```diff
@@ -40,6 +40,15 @@ var fs = require('fs'),
             alias: 'z',
             description: 'enable compression (tries to serve file of same name plus \'.gz\')'
         })
+        .option('spa', {
+            alias: 's',
+            description: 'serve the content as a single page app by redirecting all non-file requests to the index html file'
+        })
```
Comment (cloudhead) [inline bin/cli.js]: Can we drop the alias? As it's a fairly uncommon option.

File: bin/cli.js [hunk 4]
```diff
@@ -92,20 +101,33 @@ if (argv.gzip){
     (options = options || {}).gzip = true;
 }
 
+if (argv.indexFile){
```
Comment (cloudhead) [inline bin/cli.js]: Space before `{`

File: bin/cli.js [hunk 5]
```diff
@@ -92,20 +101,33 @@ if (argv.gzip){
...
+          if (e && e.status === 404) {
+              response.writeHead(e.status, e.headers);
+              response.end("Not Found");
+              log(request, response);
+          } else {
+              log(request, response);
+          }
+        }
+
+        if (argv['spa'] && !request.url.includes(".")){
```
Comment (cloudhead) [inline bin/cli.js]: Space before `{`

**Comment by cloudhead** [comment]
👍

**Comment by mmcdermott** [comment]
Thanks for the quick review! How's this look?

File: bin/cli.js [hunk 6]
```diff
@@ -65,47 +73,60 @@ var fs = require('fs'),
...
+          if (e && e.status === 404) {
+              response.writeHead(e.status, e.headers);
+              response.end("Not Found");
+              log(request, response);
+          } else {
+              log(request, response);
+          }
+        };
+
+        if (argv['spa'] && !request.url.includes(".")) {
```
Comment (jozsi) [inline bin/cli.js]: The 'includes' method is ECMAScript 2015 specific. Would require node engine >= 0.11

Comment (mmcdermott) [inline bin/cli.js]: Good catch! I've updated it to use indexOf instead. Thanks!

**Comment by cloudhead** [comment]
Great!

**Comment by preventdefault** [comment]
Hi,
the current implementation doesn´t work as expected, respectively is wrong.
If the --spa option is set, the index file will be served for all files instead for all non-files

This line
`if (argv['spa'] && request.url.indexOf(".") !== -1) {`
should be rewritten to
`if (argv['spa'] && request.url.indexOf(".") === -1) {`

And why do you use a 500 status code to serve non-files. 500 means there is a server error, but this is not the case.

Best regards
Mario

**Pull request merged**